### ISSUE TITLE
[HOLD for Payment 2024-05-29] [$250] Web - Attachment - The action bar flickers when downloading a video.

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#39598]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.60-1
**Reproducible in staging?:** y 
**Reproducible in production?:** n
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in.
3. Go to any chat
4. Add a video attachment
5. Play and Expand video using arrows in the right top corner
6. Open tree dot menu and press download

## Expected Result:
The action bar doesn't flicker when downloading a video.

## Actual Result:
he action bar flickers when downloading a video

## Workaround:
n/a

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/8c9dd3fb-11f9-47ee-a537-b6d9f7a3b1c4

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01756abc8d4e07a5e1</li>
        <li>Upwork Job ID: 1775894887205498880</li>
        <li>Last Price Increase: 2024-04-25</li>
<li>Automatic offers: </li>
<ul>
<li>rojiphil | Reviewer | 0</li>
<li>ikevin127 | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @laurenreidexpensify</details>

**Comment by kbecciv** [comment]
We think that this bug might be related to #vip-vsb

**Comment by mountiny** [comment]
This seems like a minor bug that does not effect users experience

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The video player controls blink when closing the popover menu.

### What is the root cause of that problem?
The control will show if the video player is hovered or there is a popover menu showing.
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/components/VideoPlayer/BaseVideoPlayer.tsx#L308

When we close the popover, the `isPopoverVisible` state is immediately becomes false so we can hide the popover,
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/components/VideoPlayer/BaseVideoPlayer.tsx#L326-L327

but the popover isn't actually closed yet, so the video player control is hidden for a brief moment before the video player receives the hover event back.

### What changes do you think we should make in order to solve the problem?

We need to have a new state that will tell whether the popover is really hidden or not. To do that,
1. First, we need to allow `PopoverMenu` to accept `onModalHide` prop
2. Create a new state, let's call it `isPopoverHidden` with the initial value of `true`.
3. Set it to false in `showPopoverMenu`
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/components/VideoPlayer/BaseVideoPlayer.tsx#L82-L83
4. Set it to true in `onModalHide` of the PopoverMenu.
5. Replace `isPopoverVisible` with `!isPopoverHidden`

https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/components/VideoPlayer/BaseVideoPlayer.tsx#L308

We rely on `onModalHide` too in the report context menu.
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.tsx#L186
(this is passed to the onHide above)
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/report/ReportActionItem.tsx#L332
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/report/ReportActionItem.tsx#L868

The last thing to do is to find a better name for the state, maybe `isPopoverActive`?.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

[Attachment modal]: The video player controls bar flickers when opening / closing the video's 3-dot popover menu.

### What is the root cause of that problem?

> [!tip]
> These steps are enough to reproduce the flicker:
> 1. Go to any chat.
> 2. Add a video attachment (you don't need to upload it).
> 3. Click on the 3-dot (`More`) button to open the Popover (you should see the `Playback speed` option).
> 4. Click outside of the Popover to close it -> notice the flicker.

This happens due to our [`VideoPopoverMenu`](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/VideoPlayer/BaseVideoPlayer.tsx#L326-L330) component using a [portal](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/Popover/index.tsx#L58-L75), which when **open** comes on top of the `BaseVideoPlayer` Hoverable surface of the video. These are the events that occur which lead to our issue:
1. The video player controls 3-dot menu (`More`) is clicked -> Popover opens -> `isHovered` state is set to `false`.
2. User clicks outside of the Popover -> Popover closes ->  `isHovered` state is set to `false` again.
3. Right after the Popover is gone, the Hoverable surface re-captures the mouse event -> `isHovered` state is set to `true`.

Our issue occurs between (2.) and (3.) in very quick succession which causes the flicker of the video player controls bar because of this condition [here](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/VideoPlayer/BaseVideoPlayer.tsx#L308), specifically the `isHovered` state variable coming from the `Hoverable` component which wraps the video container, toggling `false`/`true` in quick succession right after the `isPopoverVisible` becomes `false`.

### What changes do you think we should make in order to solve the problem?

For a more holistic, well-rounded solution I propose we introduce a new boolean prop to the `Hoverable` component, named `shouldFreezeCapture` which will be used within the `ActiveHoverable` component to prevent setting the `isHovered` state to `false` the second time (unnecessarily) when an open portal (popover / modal) is being closed.

**How do we use this new prop ?** We only need to apply it in two places:
1. At the top of this [`updateIsHovered`](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/Hoverable/ActiveHoverable.tsx#L19-L28) function to perform an early return if `shouldFreezeCapture` is `true`.
2. Within this useEffect's [unsetHoveredIfOutside](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/Hoverable/ActiveHoverable.tsx#L65-L71) function we pass `|| shouldFreezeCapture` to the already existing early return.

> [!note]
> Both of these are needed to prevent the unnecessary `setIsHovered(false)` state update.

This will not only fix our issue but also add a new prop which can be leveraged in the future whenever a portal (popover / modal) is open and we want to avoid the extra renders caused by the unnecessary `isHovered` state update as a result of the popover closing and re-capturing of the hover event.

### Videos (before / after)

<details>
<summary>MacOS: Chrome / Safari</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/e21e6308-da07-4429-b41e-8b5e9ed39462"> | <video src="https://github.com/Expensify/App/assets/56457735/bf18c496-958a-4efe-8819-ed9964d6f8b3"> |

</details>

**Comment by mountiny** [comment]
@rojiphil Can you please review the proposal? thanks!

**Comment by rojiphil** [comment]
I will review the proposals today and share the update.

**Comment by rojiphil** [comment]
On trying to reproduce the issue with the latest main, I could not find the flickering behavior. Instead, it simply hides the action bar (i.e. on clicking outside of the popover or on clicking on a popover menu item). This makes sense to me as the user click may occur outside or inside the video player as shown in the video below. Later, whenever there is a hover over the video player, the action bar shows up which I think is expected too. So, I am wondering if any fix is required here. 
@mountiny What are your thoughts on this? Also, attaching a test video for reference.

https://github.com/Expensify/App/assets/3004659/ece7bfa0-ff26-46e2-94d3-c5a7f2af823b

**Comment by mountiny** [comment]
That does make sense @kbecciv can you still reproduce this on latest staging?

**Comment by nkdengineer** [comment]
@rojiphil I still can reproduce the issue when playing video in "Expand" mode

**Comment by ikevin127** [comment]
@rojiphil @mountiny This issue happens within the **Attachment modal** while open with a video.

> [!tip]
> These steps are enough to reproduce the flicker:
> 1. Go to any chat.
> 2. Add a video attachment (you don't need to upload it).
> 3. Click on the 3-dot (`More`) button to open the Popover (you should see the `Playback speed` option).
> 4. Click outside of the Popover to close it -> notice the flicker.

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/e21e6308-da07-4429-b41e-8b5e9ed39462

</details>

**Comment by mountiny** [comment]
Albert, good choice

**Comment by rojiphil** [comment]
@ikevin127 @nkdengineer My earlier video also tested in the “expand” mode (i.e. within the Attachment modal). Anyway, I tried again this time while the video was still playing. And I am still not seeing the flickering. Why is it so? 

https://github.com/Expensify/App/assets/3004659/e096062e-8c8e-44e7-b6cd-53ceafea3912

**Comment by rojiphil** [comment]
@mountiny Looks like your comment [here](https://github.com/Expensify/App/issues/39598#issuecomment-2043623474) was meant for some other issue.

**Comment by rojiphil** [comment]
> You can try the above steps directly on https://staging.new.expensify.com/, but make sure to open it in an Incognito window so that there are no cached files from older builds

It looks like this was tried on Chrome as I could reproduce the problem there on the latest main. But I had tested on Safari which behaves as mentioned [here](https://github.com/Expensify/App/issues/39598#issuecomment-2043578809).  I think the behavior on Safari is what we should see on Chrome as well. Please check this on Safari so we are all on the same page.

**Comment by rojiphil** [comment]
One additional observation with the latest main is that the action bar is not at all shown in non-expanded mode on Safari. Looks like some regression somewhere. And I think we need to address this first. I have raised the issue [here](https://expensify.slack.com/archives/C049HHMV9SM/p1713117417463329).

**Comment by mountiny** [comment]
Still discussing

**Comment by mountiny** [comment]
@rojiphil what is latest on this one?

**Comment by rojiphil** [comment]
> what is latest on this one?

@mountiny The current status is that there is an inconsistency between Chrome and Safari in how the action bar behaves in the expanded mode of the video player. On Chrome the flicker occurs whereas Safari works fine as mentioned [here](https://github.com/Expensify/App/issues/39598#issuecomment-2043578809). So, the expected behavior should be that the action bar should be hidden on dismissing the popover. And, further, the action bar should be displayed based on the hover location on the video player. 

Also, it is fair to expect the same behavior in the preview mode of the video player. However, the action bar is not at all shown on my Safari. The discussion for this is happening [here](https://expensify.slack.com/archives/C049HHMV9SM/p1713117417463329). If there is an issue there, we will have to fix that first before we go ahead with fixing the flickering issue here.

**Comment by mountiny** [comment]
Thanks

**Comment by ikevin127** [comment]
> Also, it is fair to expect the same behavior in the preview mode of the video player. However, the action bar is not at all shown on my Safari. The discussion for this is happening [here](https://expensify.slack.com/archives/C049HHMV9SM/p1713117417463329). If there is an issue there, we will have to fix that first before we go ahead with fixing the flickering issue here.

After doing some digging into the Safari specific bug mentioned in Slack, with controls not appearing in non-expanded video preview I concluded that it's a Safari specific bug regarding `expo-av`'s Video component.

If we console log the variables that show the controls, we can notice in the console that it works as expected as in they toggle true false when hovering in and out of the non-expanded video component. The problem is that the non-expanded Video is freezing at some point in Safari hence nothing is shown on top of the video anymore, the only thing that can be done to resolve this situation is to click the expand button and then in expanded mode the controls show as expected.

So I think we can safely move on with fixing the flicker (this issue) and treat the Safari specific bug separately.
Any thoughts ?

cc @mountiny @rojiphil

**Comment by mountiny** [comment]
I think that sounds good. Would you be able to reproduce this bug in minimal example for the expo-av as that would help us to fix it upstream? the Expo team would definitely appreciate it

**Comment by ikevin127** [comment]
Sure, I'll invastigate that separately to make sure it's not caused by **how** we're using `expo-av`'s Video and if it's **not** on our side -> I will try and come up with a minimal example and post it upstream, maybe even come up with a patch for us in the meantime until it's fixed upstream (if the case).

I'll keep tabs here as I investigate further! (or we can create a new issue to track the progress separately)

**Comment by mountiny** [comment]
Lets just stick with it here, its related.

**Comment by ikevin127** [comment]
The other bug was just posted a few hours ago:
- https://github.com/Expensify/App/issues/40652

Some discoveries regarding the bug:
- I can reproduce with the following steps:

1.Send a video attachment.
2.Play the video so that the preview is visible.
3.Hover over the attachment preview in the report.

Now the video does play, it doesn't freeze anymore on Safari but this is what's happening with regards to video controls not showing:
- when I manually resize the safari browser window vertically (making it less taller), after a certain height point the video controls do show up normally (as expected on hover), but resizing it back to fullscreen makes the controls not show up, again.


The bug is also discussed in this [Slack 🧵](https://expensify.slack.com/archives/C049HHMV9SM/p1713117417463329), I will also post my discoveries there.

> [!note]
> It does look like there might be something on our side, either some overlaying issue which causes the video controls to not be shown in Safari, or maybe even something new arch related (if this worked as expected in Safari before new arch merge).

**Comment by mountiny** [comment]
@rojiphil can you please summarize the next steps you recommend here?

**Comment by ikevin127** [comment]
My take on https://github.com/Expensify/App/issues/40652#issuecomment-2071273071 regarding whether or not we should close #40652 in favour of this one.

**TLDR:** My take is: negative. See reasoning in comment mentioned above.

If @rojiphil cannot reproduce this one I think we should re-assign reviewer which can reproduce the issue. Otherwise I'd volunteer myself for both roles if that's possible, since according to C+ Process Doc that can be done (even though very rare).

**Comment by rojiphil** [comment]
> If [@rojiphil](https://github.com/rojiphil) cannot reproduce this one

@ikevin As mentioned [here](https://github.com/Expensify/App/issues/39598#issuecomment-2054137578), I could reproduce the flickering problem on Chrome. So reproducing the flickering problem is not the issue anymore. However, we did have a problem with reproduction of the issue of missing action bar on Safari which you were able to reproduce. Thanks for confirming on this.

**Comment by rojiphil** [comment]
> can you please summarize the next steps you recommend here?

@mountiny Continuing from [here](https://github.com/Expensify/App/issues/39598#issuecomment-2066055623), we need to fix the missing action bar issue on Safari first as we can consistently reproduce this now. I agree that both these issues are different and have different root causes. But it does not make sense to me to fix the flickering issue without fixing the missing action bar due to the simple fact that the impact on user experience due to the missing action bar issue is much larger than the flickering issue. And even if we treat both these GH issues as different, I recommend putting this on hold until we fix the missing action bar issue. This is because we need to test the flickering issue on the preview mode action bar. My recommendation though would be to fix both these issues here as we have the context here and also since they are related.

**Comment by ikevin127** [comment]
The thing is that on Safari where we have the missing video controls on non-expanded video, we don't experience the flicker (this issue) at all:
- on non-expanded video because the controls are not showing at all (#40652)
- on expanded video because this issue is not reproducible (there's no flicker on Safari)

This is why I hold my position that we do not have to hold one for the other and also we shouldn't fix both here in this issue.

cc @mountiny @rojiphil

**Comment by mountiny** [comment]
Is one of them easier to fix than another? it seems like the safari issue might be upstream, but flicker could be something easier to handle

**Comment by Skalakid** [comment]
Hi, I'm Michał from SWM, an expert agency and I will help close this issue

**Comment by Skalakid** [comment]
I think we should fix problem with flickering video action bar and missing action bar separately since they have different root causes, priority and complexity. Issue with missing action bars is far more complex and probably is connected with zIndexes and video element sharing feature. Flickering issue is less complex and we already have some proposals with the fix for it.

**Comment by Skalakid** [comment]
@rojiphil okay I will check your fix out and investigate if we can safely remove that. I remember that it was a fix for some edge case. Thanks for the info :D

**Comment by Skalakid** [comment]
Talking about flickering video action bar issue, I think the expected behavior should be to make the bar still visible after exiting the popover only when the cursor is still above the video element. If user exits the popover by clicking in other place the action bar should hide

**Comment by Skalakid** [comment]
@nkdengineer, thank you for your [proposal](https://github.com/Expensify/App/issues/39598#issuecomment-2037486004). I tested it and I found 1 bug. When you do all issue steps when the video is paused, after closing the popover, by clicking outside the video element, the action bar stops hiding when hovering out

https://github.com/Expensify/App/assets/39538890/dde9bc98-019c-4e46-9dfc-0da7a310da70

**Comment by rojiphil** [comment]
> Talking about flickering video action bar issue, I think the expected behavior should be to make the bar still visible after exiting the popover only when the cursor is still above the video element. If user exits the popover by clicking in other place the action bar should hide

Yes. That aligns with the expected behavior discussed [here](https://github.com/Expensify/App/issues/39598#issuecomment-2066055623).

**Comment by Skalakid** [comment]
@ikevin127, thank you for the proposal, I like it, and it works as expected.

> At the top of this [updateIsHovered](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/Hoverable/ActiveHoverable.tsx#L19-L28) function to perform an early return if shouldFreezeCapture is true.

The only thing that I would suggest to change is to not return true in the `updateIsHovered` function, but the same value as it was before. For example, we can freeze capturing when an element is not hovered and I'm not sure if it is a good idea to change the state to true even if nothing is hovering over it. What do you think about that?

**Comment by Skalakid** [comment]
@rojiphil I can look at the issue with video controls zIndex, however flickering action bar issue already has some proposal and we can proceed with them. For example, @ikevin127's proposal seems good to me

**Comment by rojiphil** [comment]
> I can look at the issue with video controls zIndex, however flickering action bar issue already has some proposals and we can proceed with them. For example, @ikevin127's proposal seems good to me

Sure then. Let’s align in that way. 
@ikevin127 can you please address the comment [here](https://github.com/Expensify/App/issues/39598#issuecomment-2074508645) and update the proposal to meet the expected behavior as mentioned [here](https://github.com/Expensify/App/issues/39598#issuecomment-2074395673)?

**Comment by mountiny** [comment]
Happy to assign @ikevin127 based on the convo after they address the last pending comments

**Comment by ikevin127** [comment]
@rojiphil Sure, I'll address @Skalakid's comment and update the proposal accordingly.

**Comment by ikevin127** [comment]
> I think the expected behavior should be to make the bar still visible after exiting the popover only when the cursor is still above the video element. If user exits the popover by clicking in other place the action bar should hide

@Skalakid I re-tested my [proposal](https://github.com/Expensify/App/issues/39598#issuecomment-2041187706) and this is exactly how it behaves currently so I don't think any additional changes are required.

https://github.com/Expensify/App/assets/56457735/03d234e6-f4de-40a2-aba1-b69f76725c3e

> The only thing that I would suggest to change is to not return true in the updateIsHovered function, but the same value as it was before. For example, we can freeze capturing when an element is not hovered and I'm not sure if it is a good idea to change the state to true even if nothing is hovering over it. What do you think about that?

Not sure I'm following here, given my [proposal](https://github.com/Expensify/App/issues/39598#issuecomment-2041187706) suggests early return if `shouldFreezeCapture` is true, then the `setIsHovered` state update at the end of the `updateIsHovered` function is never reached hence maintaining (freezing) the current `isHovered`, whatever that is.

Again, as explained above regarding your first quoted request, I don't see the need to add any extra logic to the proposed solution considering that it already behaves as required.

If I'm missing anything, please let me know and I can update the proposal - otherwise, if all parts agree with my assignment, I can open the PR today!

**Comment by Skalakid** [comment]
@ikevin127 that makes sense, sorry I probably misunderstood something. Thanks for explanation

**Comment by mountiny** [comment]
I will go ahead and assign @ikevin127 as he has shown deep understanding of this issue

**Comment by ikevin127** [comment]
PR https://github.com/Expensify/App/pull/41125 ready for review! 🎉

**Comment by rojiphil** [comment]
Thanks @ikevin127 for the PR and the relevant updates. 
I will review and share the update today.

**Comment by rojiphil** [comment]
PR deployed to staging. Awaiting production deployment.

**Comment by ikevin127** [comment]
⚠️ **Automation failed**: This was deployed to **production** today [here](https://github.com/Expensify/App/pull/41125#issuecomment-2125961535), therefore should be **[HOLD for Payment 2024-05-29]** and have `Awaiting Payment` label. Additionally, it's missing the `Bug` label -> doesn't have a BZ team member assigned.

cc @mountiny

**Comment by mountiny** [comment]
$250 to @ikevin127 and to @rojiphil

**Comment by laurenreidexpensify** [comment]
Payment only due next week

**Comment by laurenreidexpensify** [comment]
Payment Summary:

- C+ @rojiphil $250 issued in upwork 
- Contributor @ikevin127 $250 issued in upwork 

Looks like this was reported by Applause, no regression test required. Closing
